File: Source/TagCounter.cs

namespace NhlStenden.TagBag;

/// <summary>
/// Represents a class that counts the occurrences of tags in files.
/// </summary>
public class TagCounter : TagBase
{
    /// <summary>
    /// Executes the tag counting process on the specified files.
    /// </summary>
    /// <param name="args">A list of arguments where the first argument is the directory path.</param>
    public static void Execute(IList<string> args)
    {
        var fileNames = GetTagFiles(args[0]);
        var tokens = new Dictionary<string, int>();
        CountTagsInFiles(fileNames, tokens);

        foreach (var token in tokens.OrderByDescending(x => x.Value))
            Console.WriteLine(token.Key + ": " + token.Value);
    }

    /// <summary>
    /// Counts the tags in the specified files and updates the token dictionary.
    /// </summary>
    /// <param name="fileNames">An array of file names to process.</param>
    /// <param name="tokens">A dictionary to store the count of each tag.</param>
    static void CountTagsInFiles(string[] fileNames, Dictionary<string, int> tokens)
    {
        foreach (var fileName in fileNames)
            CountTagsInFile(fileName, tokens);
    }

    /// <summary>
    /// Counts the tags in the specified file and updates the token dictionary.
    /// </summary>
    /// <param name="fileName">The name of the file to process.</param>
    /// <param name="tokens">A dictionary to store the count of each tag.</param>
    static void CountTagsInFile(string fileName, IDictionary<string, int> tokens)
    {
        var lines = File.ReadAllLines(fileName).ToList();

        for (var i = 0; i < lines.Count; i++)
        {
            var line = lines[i];

            var items = TagHelper.SplitIntoTags(line);
            for (var j = 0; j < items.Count; j++)
            {
                var item = items[j];
                if (tokens.TryGetValue(item, out var counter))
                {
                    tokens[item] = counter + 1;
                }
                else
                {
                    tokens[item] = 1;
                }
            }
        }
    }    
}

File: tagbag/tag_counter.py
from .tag_base import TagBase
from .tag_helper import TagHelper


from pathlib import Path
from typing import Dict, Iterable, Sequence


class TagCounter(TagBase):
    """
    Count frequency of tags across all files and print counts to stdout.
    """

    @staticmethod
    def execute(args: Sequence[str]) -> None:
        """
        args[0]: directory path
        """
        file_names = TagBase.get_tag_files(args[0])
        tokens: Dict[str, int] = {}
        TagCounter._count_tags_in_files(file_names, tokens)

        for token, count in sorted(tokens.items(), key=lambda kv: kv[1], reverse=True):
            print(f"{token}: {count}")

    @staticmethod
    def _count_tags_in_files(file_names: Iterable[str], tokens: Dict[str, int]) -> None:
        for name in file_names:
            TagCounter._count_tags_in_file(Path(name), tokens)

    @staticmethod
    def _count_tags_in_file(file_path: Path, tokens: Dict[str, int]) -> None:
        for line in file_path.read_text(encoding="utf-8").splitlines():
            for item in TagHelper.split_into_tags(line):
                tag = item.strip()
                if not tag:
                    continue
                tokens[tag] = tokens.get(tag, 0) + 1